client/app/visualizations/choropleth/index.js

```diff
@@ -0,0 +1,265 @@
+        if (map) {
+          map.remove();
+        }
+        if (!_.isObject(countriesData)) {
+          return;
+        }
+
+        $scope.formatValue = createNumberFormatter(
+          $scope.options.valueFormat,
+          $scope.options.noValuePlaceholder,
+        );
+
+        const data = prepareData(
+          $scope.queryResult.getData(),
+          $scope.options.countryCodeColumn,
+          $scope.options.valueColumn,
+        );
+
+        const { limits, colors, legend } = createScale(
+          countriesData.features,
+          data,
+          $scope.options,
+        );
+
+        // Update data for legend block
+        $scope.legendItems = legend;
+
+        choropleth = L.geoJson(countriesData, {
+          onEachFeature: (feature, layer) => {
+            const value = getValueForFeature(feature, data, $scope.options.countryCodeType);
+            const valueFormatted = $scope.formatValue(value);
+            const featureData = prepareFeatureProperties(
+              feature,
+              valueFormatted,
+              data,
+              $scope.options.countryCodeType,
+            );
+            const color = getColorByValue(value, limits, colors, $scope.options.colors.noValue);
+
+            layer.setStyle({
+              color: $scope.options.colors.borders,
+              weight: 1,
+              fillColor: color,
+              fillOpacity: 1,
+            });
+
+            if ($scope.options.tooltip.enabled) {
+              layer.bindTooltip($sanitize(formatSimpleTemplate(
+                $scope.options.tooltip.template,
+                featureData,
+              )));
+            }
+
+            if ($scope.options.popup.enabled) {
+              layer.bindPopup($sanitize(formatSimpleTemplate(
+                $scope.options.popup.template,
+                featureData,
+              )));
+            }
+
+            layer.on('mouseover', () => {
+              layer.setStyle({
+                weight: 2,
+                fillColor: darkenColor(color),
+              });
+            });
+            layer.on('mouseout', () => {
+              layer.setStyle({
+                weight: 1,
+                fillColor: color,
+              });
+            });
+          },
+        });
+
+        const choroplethBounds = choropleth.getBounds();
+
+        map = L.map($element[0].children[0].children[0], {
+          center: choroplethBounds.getCenter(),
+          zoom: 1,
+          layers: [choropleth],
+          scrollWheelZoom: false,
+          maxBounds: choroplethBounds,
+          maxBoundsViscosity: 1,
+          attributionControl: false,
+        });
+      }
+
+      function resize() {
+        if (map) {
+          map.invalidateSize(false);
+          map.fitBounds(choropleth.getBounds());
+        }
+      }
+
+      $scope.handleResize = () => {
+        resize();
+      };
+
+      $http.get(countriesDataUrl).then((response) => {
+        countriesData = response.data;
```

Comment:
How about we wrap loading `countriesData` in a module global function that returns a promise? This way it will be cached while the app is loaded.

Also it solves the issue of `render` being called before the data is ready: `render` can wait for this promise to resolve (which is immediate if it's already resolved).